chain/sync.go

```diff
@@ -478,43 +467,97 @@ func (syncer *Syncer) ValidateBlock(ctx context.Context, b *types.FullBlock) err
-		return xerrors.Errorf("validating block tickets failed: %w", err)
-	}
+		waddr, err := stmgr.GetMinerWorker(ctx, syncer.sm, stateroot, h.Miner)
+		if err != nil {
+			return nil, xerrors.Errorf("GetMinerWorker failed: %w", err)
+		}
+		return waddr, nil
+	})
+
+	minerCheck := async.Err(func() error {
+		if err := syncer.minerIsValid(ctx, h.Miner, baseTs); err != nil {
+			return xerrors.Errorf("minerIsValid failed: %w", err)
+		}
+		return nil
+	})
 
-	rand, err := syncer.sm.ChainStore().GetRandomness(ctx, baseTs.Cids(), h.Tickets, build.RandomnessLookback)
+	waddri, err := waddrPromise.AwaitContext(ctx)
 	if err != nil {
-		return xerrors.Errorf("failed to get randomness for verifying election proof: %w", err)
+		return err
 	}
+	waddr := waddri.(address.Address)
 
-	if err := VerifyElectionProof(ctx, h.ElectionProof, rand, waddr); err != nil {
-		return xerrors.Errorf("checking eproof failed: %w", err)
-	}
+	blockSigCheck := async.Err(func() error {
+		if err := h.CheckBlockSignature(ctx, waddr); err != nil {
+			return xerrors.Errorf("check block signature failed: %w", err)
+		}
+		return nil
+	})
 
-	mpow, tpow, err := stmgr.GetPower(ctx, syncer.sm, baseTs, h.Miner)
-	if err != nil {
-		return xerrors.Errorf("failed getting power: %w", err)
-	}
+	tktsCheck := async.Err(func() error {
+		if err := syncer.validateTickets(ctx, waddr, h.Tickets, baseTs); err != nil {
+			return xerrors.Errorf("validating block tickets failed: %w", err)
+		}
+		return nil
+	})
 
-	if !types.PowerCmp(h.ElectionProof, mpow, tpow) {
-		return xerrors.Errorf("miner created a block but was not a winner")
-	}
+	eproofCheck := async.Err(func() error {
+		rand, err := syncer.sm.ChainStore().GetRandomness(ctx, baseTs.Cids(), h.Tickets, build.RandomnessLookback)
+		if err != nil {
+			return xerrors.Errorf("failed to get randomness for verifying election proof: %w", err)
+		}
 
-	if err := syncer.checkBlockMessages(ctx, b, baseTs); err != nil {
-		return xerrors.Errorf("block had invalid messages: %w", err)
+		if err := VerifyElectionProof(ctx, h.ElectionProof, rand, waddr); err != nil {
+			return xerrors.Errorf("checking eproof failed: %w", err)
+		}
+		return nil
+	})
+
+	winnerCheck := async.Err(func() error {
+		mpow, tpow, err := stmgr.GetPower(ctx, syncer.sm, baseTs, h.Miner)
+		if err != nil {
+			return xerrors.Errorf("failed getting power: %w", err)
+		}
+
+		if !types.PowerCmp(h.ElectionProof, mpow, tpow) {
+			return xerrors.Errorf("miner created a block but was not a winner")
+		}
+		return nil
+	})
+
+	msgsCheck := async.Err(func() error {
+		if err := syncer.checkBlockMessages(ctx, b, baseTs); err != nil {
+			return xerrors.Errorf("block had invalid messages: %w", err)
+		}
+		return nil
+	})
+
+	await := []async.ErrorFuture{
+		minerCheck,
+		tktsCheck,
+		blockSigCheck,
+		eproofCheck,
+		winnerCheck,
+		msgsCheck,
+	}
+	for _, fut := range await {
+		if err := fut.AwaitContext(ctx); err != nil {
+			return err
```

Comment:
Only thing I'm slightly uncomfortable with here is that it might hide errors since it drops everything after the first error it encounters